### PR TITLE
Add withRoute Higher-Order Component

### PR DESCRIPTION
**Pull request by HaNdTriX**
I did this PR based on the awesome work of @nwshane in #2056.
Since in this PR the api differs from the initial proposal I decided to create a new PR.

## Differences to #2056

* This PR only exposes the current route instead of the entire router. A route is an Object with the following props: `pathname`, `query`, `asPath`. Later we may add more props if we really need them (e.g.: `components`, `query`).
* ~The `withRoute` hoc takes a reducer function instead of an `options` object. Similar to redux connect method.~
* Name `withRoute` instead of `withRouter`

## API

`withRoute(<Component>)`

~The `mapRouterToProps` function is optional. If you omit it you will get the route object with the following properties:~

**Route object:**

- `pathname` - path section of URL
- `query` - query string section of URL parsed as an object
- `asPath` - the actual url path

## Example

```javascript
import { withRoute } from 'next/router'

export default withRoute(LinkComponent)
```

[examples/using-router](https://github.com/HaNdTriX/next.js/blob/with-route/examples/using-router)

## Todo

* [x] Add route context object
* [x] Add HOC (`withRoute`)
* [x] Add [documentation](https://github.com/HaNdTriX/next.js/tree/with-route#withroute-higher-order-component)
* [x] Update [using-router Example](https://github.com/HaNdTriX/next.js/tree/with-route/examples/using-router)
* [ ] Add tests

### CODE REVIEW

**Comment by HaNdTriX** [comment]
@fridays does it makes sense to somehow change the API to improve integration with [next-routes](https://github.com/fridays/next-routes)?

**Comment by nwshane** [comment]
@HaNdTriX Interesting idea! I like that it limits the users' access to the Router properties they need. 

Why did you go with using a `mapRouteToProps` function, instead of just passing `route` as a prop into the wrapped component?

My knowledge of the code base and overall Next JS API isn't strong enough to comment on whether this or the withRouter HOC would be a better approach - I'll let people with more experience discuss that.

Good job!

**Comment by timneutkens** [comment]
This looks really cool. I don't like the `mapRouteToProps` though. We could just pass down `route` and the user could implement their own `mapRouteToProps` type stuff in a HOC.

**Comment by timneutkens** [comment]
@arunoda what are your thoughts on this?

**Comment by HaNdTriX** [comment]
**Status**: Waiting for feedback.

**Comment by arunoda** [comment]
Guys, this is awesome.
I was waiting to release 3.0 and start working on these and take all.
This is definitely going into the core after 3.0

**Comment by HaNdTriX** [comment]
@timneutkens & @nwshane thanks for the awesome feedback! I totally agree that `mapRouteToProps` is not needed, since a dev could aways implement this using something like [recompose](https://github.com/acdlite/recompose) to add this kind of functionality.

[Removed `mapRouteToProps`](https://github.com/zeit/next.js/pull/2352/commits/90343a228963322d2230a7479694f6ee425616a5)

@arunoda Thanks for having a look at this.

**Comment by timneutkens** [comment]
@HaNdTriX 👌 🙌

**Comment by fridays** [comment]
Looks great @HaNdTriX! 👍

**Comment by timneutkens** [comment]
I'll try to add some tests to the PR this week. So we can take it 👌

**Comment by HaNdTriX** [comment]
@timneutkens I just rebased/squashed this branch with the current master in order to keep your git history clean.

## Changelog

- squashed commit: removed routerToProps
- updated hoist-non-react-statics to version 2
- improved propTypes

**Comment by timneutkens** [comment]
Thanks @HaNdTriX ❤️

**Comment by jedlikowski** [comment]
Can this be merged? 😊

**Comment by evercool** [comment]
Pls merge this. All workarounds I've found or think of are weird or doesn't work at all.

**Comment by hurvajs77** [comment]
This pull is really important for us work. We left the Este project, but we're very disappointed by state of Next.js... We hope that Next will improve early

**Comment by HaNdTriX** [comment]
## TODO

* [ ] tests

In addition to that I believe it's a good thing that it hasn't been merged. This just means that the guys at zeit are thinking a lot about the fact that there might be better ways to solve this problem 👏 . 

So from my point of view. @timneutkens take your time. Evaluate if there might be better ways. I'll be fine with that. Actually I really enjoy next since there haven't been a lot of breaking changes.

For the others hack it for now and replace it when this PR got merged:
To point you in a potential direction check out the following code:

```js
import Router from 'next/router'
// Keep in mind that this API might not work in 
// future releases of next.js

function isActive(href) {
  if(!process.browser) {
    // Router is clientOnly :(
    return false
  }
  return Router.route === href
}
```

File: lib/app.js [hunk 1]
```diff
@@ -2,15 +2,18 @@ import React, { Component } from 'react'
 import PropTypes from 'prop-types'
```
Comment (timneutkens) [inline lib/app.js]: @HaNdTriX so what I was thinking. Why don't we just expose the context, and create a separate package that reads the context, that way we don't really have to add any code to Next, but provide a very powerful api 👌

Comment (timneutkens) [inline lib/app.js]: In terms of use it wouldn't differ much. But it allows us to not hog in another library / extra code to handle the withRoute case which I'm pretty sure not everyone is going to use 👍

Comment (timneutkens) [inline lib/app.js]: I'd like to experiment with that solution

**Comment by evercool** [comment]
If I might be so bold since Im new to react, I would like to share my workaround: https://github.com/evercool/nextjs-link
It uses HOCs wrapping page component (you would probably use some kind of HOC anyway like withRedux), which takes all props from getInitialProps of page and pass pathname into context. Link component just consume this context's property.

**Comment by timneutkens** [comment]
@evercool 💯 Nice I thought about that one too 👌 This solution uses redux etc though, so I'll draft something that's solution independent 👍

**Comment by idhard** [comment]
could you please  tell us what is the final resolution to this matter ? is this pull request going to be merged ? could we base our job on it ? if not would you please indicate me which may be the prefered way to handle the active links in a component ? thanks

**Comment by por** [comment]
@idhard I used @evercool’s solution (without the Redux part) and I think it’s an elegant solution.

```js
// hocs/with-context.js

import { Component } from 'react';
import PropTypes from 'prop-types';

export default (ComposedComponent) =>
  class WithContext extends Component {
    static childContextTypes = {
      pathname: PropTypes.string,
    };
    getChildContext() {
      return {
        pathname: this.props.url.pathname,
      };
    }

    render() {
      return <ComposedComponent {...this.props} />;
    }
  };
```

Use the hoc like `export default withContext(PageView)` in your page and child components will then have access to `this.context.pathname`. Don’t forget to specify `contextTypes` in the child view like:

```js
ActiveLink.contextTypes = {
  pathname: PropTypes.string,
};
```

**Comment by evercool** [comment]
@idhard I left withRedux in my code because I thought everybody uses redux or something like that and I wanted to keep a hint how to embed another HOCs like Intl, Authorization...

**Comment by idhard** [comment]
@por , @evercool  thanks for the example , i will definitely come out with something using it as base , however still is not clear for me what is the position of Next team about it @timneutkens

**Comment by timneutkens** [comment]
@idhard made an official solution yesterday evening. Will release it tonight 👌

**Comment by idhard** [comment]
@timneutkens  cool! looking forward to it . thanks

**Comment by arunoda** [comment]
This PR is very well written. I'm going to write some tests and take this in.

**Comment by arunoda** [comment]
@HaNdTriX thanks for the great work on this PR.
I'm moving this into https://github.com/zeit/next.js/pull/2870 just to make some few changes and take this in.

**Comment by arunoda** [comment]
I just published [3.2.0](https://github.com/zeit/next.js/releases/tag/3.2.0) which has this PR.

**Pull request closed without merging**